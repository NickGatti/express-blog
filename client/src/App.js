import React from 'react';
import PropTypes from 'prop-types';
import baseStyle from './style.scss';
import Blog from './BlogHome';
import Login from './Login';
import NewPost from './NewPost';

const link = (text, onClick) => (
  <a
    href="#"
    onClick={e => {
      e.preventDefault();
      onClick();
    }}
  >
    {text}
  </a>
);

const App = props => (
  <div>
    <h1 className={baseStyle.headerBorder}>Blog</h1>
    <div>
      <ul>
        <li>
          {props.app.page === 'blog'
            ? 'Home'
            : link('Home', props.action.appAction.gotoBlog)}
        </li>
        {props.app.loggedIn ? (
          <li>
            {props.app.page === 'newPost'
              ? 'New Post'
              : link('New Post', props.action.appAction.gotoNewPost)}
          </li>
        ) : (
          <li>
            {props.app.page === 'login'
              ? 'Login'
              : link('Login', props.action.appAction.gotoLogin)}
          </li>
        )}
      </ul>
    </div>
    <div>
      {(() => {
        switch (props.app.page) {
          case 'blog':
            return <Blog blog={props.blog} action={props.action.blogAction} />;
          case 'login':
            return (
              <Login
                login={props.login}
                loginAction={props.action.loginAction}
              />
            );
          case 'newPost':
            return (
              <NewPost
                newPost={props.newPost}
                newPostAction={props.action.newPostAction}
              />
            );
        }
      })()}
    </div>
  </div>
);
App.propTypes = {
  app: PropTypes.shape({
    page: PropTypes.string.isRequired,
    pageState: PropTypes.any,
    loggedIn: PropTypes.bool.isRequired,
  }).isRequired,
  blog: PropTypes.shape({
    posts: PropTypes.objectOf(
      PropTypes.shape({
        title: PropTypes.string.isRequired,
        post: PropTypes.string.isRequired,
      })
    ).isRequired,
    list: PropTypes.arrayOf(PropTypes.number),
    loadingPosts: PropTypes.bool.isRequired,
    error: PropTypes.string,
  }),
  action: PropTypes.shape({
    appAction: PropTypes.shape({
      gotoBlog: PropTypes.func.isRequired,
      gotoNewPost: PropTypes.func.isRequired,
      gotoLogin: PropTypes.func.isRequired,
    }).isRequired,
    blogAction: PropTypes.object.isRequired,
  }),
};

export default App;
