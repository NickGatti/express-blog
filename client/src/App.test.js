/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

describe('The App component', () => {
  const baseProps = {
    app: {
      page: 'blog',
      pageState: 1,
      loggedIn: false,
    },
    blog: {
      posts: {},
      loadingPosts: true,
      list: [],
      error: null,
    },
    action:{
      appAction: {
        gotoBlog: jest.fn(),
        gotoNewPost: jest.fn(),
        gotoLogin: jest.fn(),
      },
      blogAction: {
        loadPosts: jest.fn(),
        loadPost: jest.fn(),
      },
      loginAction: {
        logout: jest.fn(),
      },
    },
  };

  test('Should render initial view while waiting for data', () => {
    const props = baseProps;
    const renderedApp = renderer.create(<App {...props} />);
    expect(renderedApp.toJSON()).toMatchSnapshot();
  });
  test('Should render blog posts', () => {
    const props = {
      ...baseProps,
      blog: {
        posts: {
          '1': {
            id: 1,
            title: 'Post about nonsense',
            author: 1,
            created_at: '2017-10-19T17:59:42.795Z',
            updated_at: '2017-10-19T17:59:42.795Z',
            post: 'This is a short demo post about nothing.',
          },
          '2': {
            id: 2,
            title: 'React Basic Input',
            author: 1,
            created_at: '2017-10-24T19:16:55.722Z',
            updated_at: '2017-10-24T19:16:55.722Z',
            post: 'This is the second post now! I wrote a thing about a thing!',
          },
        },
        loadingPosts: false,
        list: [2,1],
        error: null,
      },
    };
    const renderedApp = renderer.create(<App {...props} />);
    expect(renderedApp.toJSON()).toMatchSnapshot();
  });
});
