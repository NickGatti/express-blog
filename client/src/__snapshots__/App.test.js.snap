// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App component Should render blog posts 1`] = `
<div>
  <h1>
    Blog
  </h1>
  <div>
    <ul>
      <li>
        Home
      </li>
      <li>
        <a
          href="#"
          onClick={[Function]}
        >
          Login
        </a>
      </li>
    </ul>
  </div>
  <div>
    <div>
      <div>
        <div>
          <h2>
            React Basic Input
          </h2>
          <div>
            This is the second post now! I wrote a thing about a thing!
          </div>
        </div>
        <div>
          <h2>
            Post about nonsense
          </h2>
          <div>
            This is a short demo post about nothing.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`The App component Should render initial view while waiting for data 1`] = `
<div>
  <h1>
    Blog
  </h1>
  <div>
    <ul>
      <li>
        Home
      </li>
      <li>
        <a
          href="#"
          onClick={[Function]}
        >
          Login
        </a>
      </li>
    </ul>
  </div>
  <div>
    <div>
      <div>
        Loading posts...
      </div>
    </div>
  </div>
</div>
`;
